experimental/ir/fdp.go

```diff
@@ -0,0 +1,443 @@
+// Copyright 2020-2025 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ir
+
+import (
+	"strings"
+
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/descriptorpb"
+
+	"github.com/bufbuild/protocompile/experimental/ast/predeclared"
+	"github.com/bufbuild/protocompile/experimental/ast/syntax"
+	"github.com/bufbuild/protocompile/experimental/ir/presence"
+	"github.com/bufbuild/protocompile/experimental/seq"
+	"github.com/bufbuild/protocompile/internal/ext/slicesx"
+	"github.com/bufbuild/protocompile/internal/intern"
+	"github.com/bufbuild/protocompile/wellknownimports"
+)
+
+// FDS code-generates a FileDescriptorSet for the given files, and returns the
+// result as an encoded byte slice.
+//
+// The resulting FDS is always fully linked: it contains all dependencies except
+// the WKTs, and all names are fully-qualified.
+func FDS(files []File, options ...CodegenOption) ([]byte, error) {
+	var c codegen
+	for _, opt := range options {
+		if opt != nil {
+			opt(&c)
+		}
+	}
+
+	fds := new(descriptorpb.FileDescriptorSet)
+	c.files(files, fds)
+	return proto.Marshal(fds)
+}
+
+// FDP code-generates a single FileDescriptorProto for file, and returns the
+// result as an encoded byte slice.
+//
+// The resulting FDP is fully linked: all names are fully-qualified.
+func FDP(file File, options ...CodegenOption) ([]byte, error) {
+	var c codegen
+	for _, opt := range options {
+		if opt != nil {
+			opt(&c)
+		}
+	}
+
+	fdp := new(descriptorpb.FileDescriptorProto)
+	c.file(file, fdp)
+	return proto.Marshal(fdp)
+}
+
+// CodegenOption is an option to pass to [DescriptorSet].
+type CodegenOption func(*codegen)
+
+// IncludeWKTs is an option for including FDPs for the Well-Known Types in
+// codegen output.
+func IncludeWKTs() CodegenOption {
+	return func(c *codegen) { c.includeWKTs = true }
+}
+
+// IncludeSourceCodeInfo is an option that requires generating source code info
+// for the descriptors.
+func IncludeSourceCodeInfo() CodegenOption {
+	return func(c *codegen) { c.includeSourceCodeInfo = true }
+}
+
+type codegen struct {
+	includeWKTs           bool
+	includeSourceCodeInfo bool
+
+	currentFile File
+}
+
+func (c *codegen) files(files []File, fds *descriptorpb.FileDescriptorSet) {
+	if c.includeSourceCodeInfo {
+		panic("protocompile/ir: NYI: IncludeSourceCodeInfo()")
```

Comment:
I think it would be more straight-forward to just omit the unimplemented options for now and replace them with TODOs to add them later. If you insist on keeping them and having them result in panics, I think we can at least spell out the error -- "NYI" is not a very common acronym, at least not one I've ever seen before.

Reply:
Works for me.